File: plugins/wise/engine/src/scheduler.ts
// DAG scheduler: wave computation from `depends_on`, trigger rules, `when:` gating,
// skip propagation and run-completion detection. Port of `_step_by_id`,
// `_trigger_rule_satisfied` and `cmd_next_wave` from scripts/workflows.py, plus the
// v2 `when:` expression evaluator. Pure: no I/O.

import type { State, StepState, Step, WorkflowDef } from "./types.ts";
import { TERMINAL_STEP, TRIGGER_RULES } from "./types.ts";

// ---- trigger rules ----------------------------------------------------------

export type TriggerVerdict = { runnable: boolean; skip: boolean };

type StatusOnly = Pick<StepState, "status">;

export function stepById(steps: readonly Step[], id: string): Step | undefined {
  return steps.find((s) => s.id === id);
}

/** Return whether the step may run now and whether it must be skipped instead. */
export function triggerRuleSatisfied(rule: string, deps: readonly StatusOnly[]): TriggerVerdict {
  if (deps.length === 0) return { runnable: true, skip: false };
  const statuses = deps.map((d) => d.status);
  const terminal = statuses.filter((s) => TERMINAL_STEP.has(s)).length;
  const completed = statuses.filter((s) => s === "completed").length;
  const failed = statuses.filter((s) => s === "failed").length;
  const allTerminal = terminal === statuses.length;

  switch (rule) {
    case "all-success":
      if (completed === statuses.length) return { runnable: true, skip: false };
      if (statuses.some((s) => s === "failed" || s === "skipped" || s === "cancelled")) {
        return { runnable: false, skip: true };
      }
      return { runnable: false, skip: false };
    case "one-success":
      if (completed > 0) return { runnable: true, skip: false };
      if (allTerminal) return { runnable: false, skip: true };
      return { runnable: false, skip: false };
    case "all-done":
      return { runnable: allTerminal, skip: false };
    case "none-failed":
      // Tolerates ALL deps skipped (user-deselected stages); a failed dep
      // propagates the skip before the rest finish.
      if (failed > 0) return { runnable: false, skip: true };
      return { runnable: allTerminal, skip: false };
    case "none-failed-min-one-success":
      if (allTerminal && failed === 0 && completed > 0) return { runnable: true, skip: false };
      if (failed > 0) return { runnable: false, skip: true };
      return { runnable: false, skip: false };
    default:
      // Unknown rule behaves like all-success.
      return triggerRuleSatisfied("all-success", deps);
  }
}

export function isTriggerRule(rule: string): boolean {
  return (TRIGGER_RULES as readonly string[]).includes(rule);
}

// ---- `when:` expression evaluator ---------------------------------------------
//
// Grammar (lowest to highest precedence):
//   or     := and ('||' and)*
//   and    := eq ('&&' eq)*
//   eq     := unary (('==' | '!=') unary)*
//   unary  := '!' unary | primary
//   primary := '(' or ')' | string | number | true | false | identifier
// Identifiers may contain dots. Unset identifiers evaluate to `undefined`.

type TokKind = "ident" | "string" | "number" | "op" | "lparen" | "rparen" | "end";
type Token = { kind: TokKind; value: string; pos: number };

function whenError(expr: string, message: string, pos: number): Error {
  return new Error(`when: ${message} at position ${pos} in ${JSON.stringify(expr)}`);
}

function isIdentStart(ch: string): boolean {
  return /[A-Za-z_]/.test(ch);
}

function isIdentPart(ch: string): boolean {
  return /[A-Za-z0-9_]/.test(ch);
}

function tokenize(expr: string): Token[] {
  const tokens: Token[] = [];
  let i = 0;
  while (i < expr.length) {
    const ch = expr[i] as string;
    if (/\s/.test(ch)) {
      i += 1;
      continue;
    }
    const start = i;
    if (ch === "(") {
      tokens.push({ kind: "lparen", value: ch, pos: start });
      i += 1;
    } else if (ch === ")") {
      tokens.push({ kind: "rparen", value: ch, pos: start });
      i += 1;
    } else if (ch === "'" || ch === '"') {
      let j = i + 1;
      while (j < expr.length && expr[j] !== ch) j += 1;
      if (j >= expr.length) throw whenError(expr, "unterminated string", start);
      tokens.push({ kind: "string", value: expr.slice(i + 1, j), pos: start });
      i = j + 1;
    } else if (/[0-9]/.test(ch)) {
      let j = i;
      while (j < expr.length && /[0-9.]/.test(expr[j] as string)) j += 1;
      const text = expr.slice(i, j);
      if (!/^\d+(\.\d+)?$/.test(text)) throw whenError(expr, `bad number ${text}`, start);
      tokens.push({ kind: "number", value: text, pos: start });
      i = j;
    } else if (isIdentStart(ch)) {
      let j = i;
      while (j < expr.length && (isIdentPart(expr[j] as string) || expr[j] === ".")) j += 1;
      const text = expr.slice(i, j);
      if (!/^[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)*$/.test(text)) {
        throw whenError(expr, `bad identifier ${text}`, start);
      }
      tokens.push({ kind: "ident", value: text, pos: start });
      i = j;
    } else {
      const two = expr.slice(i, i + 2);
      if (two === "==" || two === "!=" || two === "&&" || two === "||") {
        tokens.push({ kind: "op", value: two, pos: start });
        i += 2;
      } else if (ch === "!") {
        tokens.push({ kind: "op", value: ch, pos: start });
        i += 1;
      } else {
        throw whenError(expr, `unexpected character ${JSON.stringify(ch)}`, start);
      }
    }
  }
  tokens.push({ kind: "end", value: "", pos: expr.length });
  return tokens;
}

/** Bare identifier truthiness: non-empty string, non-zero number, true, non-empty array. */
export function truthy(value: unknown): boolean {
  if (value === undefined || value === null) return false;
  if (typeof value === "string") return value !== "";
  if (typeof value === "number") return value !== 0 && !Number.isNaN(value);
  if (typeof value === "boolean") return value;
  if (Array.isArray(value)) return value.length > 0;
  return true;
}

/** Unset never equals a set value; numbers and booleans compare to strings by their text. */
function valuesEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === undefined || b === undefined || a === null || b === null) return false;
  if (typeof a === "object" || typeof b === "object") return false;
  return String(a) === String(b);
}

function walk(root: unknown, path: readonly string[]): unknown {
  let cur: unknown = root;
  for (const seg of path) {
    if (cur === null || typeof cur !== "object") return undefined;
    cur = (cur as Record<string, unknown>)[seg];
  }
  return cur;
}

const BARE_ROOTS = ["outputs", "inputs", "answers"] as const;

/** Dotted names walk from the scope root; unknown roots fall back to outputs, inputs, answers. */
export function resolveIdentifier(name: string, scope: Record<string, unknown>): unknown {
  const path = name.split(".");
  const head = path[0] as string;
  if (Object.hasOwn(scope, head)) return walk(scope, path);
  for (const root of BARE_ROOTS) {
    const bucket = scope[root];
    if (bucket !== null && typeof bucket === "object" && Object.hasOwn(bucket, head)) {
      return walk(bucket, path);
    }
  }
  return undefined;
}

export function evaluateWhen(expr: string, scope: Record<string, unknown>): boolean {
  const tokens = tokenize(expr);
  let idx = 0;
  const peek = (): Token => tokens[idx] as Token;
  const next = (): Token => tokens[idx++] as Token;
  const isOp = (value: string): boolean => peek().kind === "op" && peek().value === value;

  const parsePrimary = (): unknown => {
    const tok = next();
    switch (tok.kind) {
      case "lparen": {
        const value = parseOr();
        if (peek().kind !== "rparen") throw whenError(expr, "expected ')'", peek().pos);
        idx += 1;
        return value;
      }
      case "string":
        return tok.value;
      case "number":
        return Number(tok.value);
      case "ident":
        if (tok.value === "true") return true;
        if (tok.value === "false") return false;
        return resolveIdentifier(tok.value, scope);
      case "end":
        throw whenError(expr, "unexpected end of expression", tok.pos);
      default:
        throw whenError(expr, `unexpected ${JSON.stringify(tok.value)}`, tok.pos);
    }
  };

  const parseUnary = (): unknown => {
    if (isOp("!")) {
      idx += 1;
      return !truthy(parseUnary());
    }
    return parsePrimary();
  };

  const parseEq = (): unknown => {
    let left = parseUnary();
    while (isOp("==") || isOp("!=")) {
      const op = next().value;
      const right = parseUnary();
      const eq = valuesEqual(left, right);
      left = op === "==" ? eq : !eq;
    }
    return left;
  };

  const parseAnd = (): unknown => {
    let left = parseEq();
    while (isOp("&&")) {
      idx += 1;
      const right = parseEq();
      left = truthy(left) && truthy(right);
    }
    return left;
  };

  const parseOr = (): unknown => {
    let left = parseAnd();
    while (isOp("||")) {
      idx += 1;
      const right = parseAnd();
      left = truthy(left) || truthy(right);
    }
    return left;
  };

  const result = parseOr();
  const tail = peek();
  if (tail.kind !== "end") {
    throw whenError(expr, `unexpected ${JSON.stringify(tail.value)}`, tail.pos);
  }
  return truthy(result);
}

// ---- next wave ----------------------------------------------------------------

export type SkipReason = { id: string; reason: string };
export type Wave = {
  ready: Step[];
  skipped: SkipReason[];
  /** No step can run or be skipped and nothing is in flight. */
  done: boolean;
  /** Set with `done` when a step failed or a pending step is unreachable. */
  failed: boolean;
  /** Non-fatal notices, e.g. an unparseable `when:` treated as true. */
  warnings: string[];
};

/** The scope `when:` expressions evaluate against. */
export function whenScope(state: State): Record<string, unknown> {
  return { outputs: state.outputs, inputs: state.inputs, answers: state.answers };
}

/** `when:` is a string in v2; a v1 list of conditions is still accepted and AND-ed. */
function whenConditions(when: unknown): string[] {
  if (Array.isArray(when)) return when.map((c) => String(c));
  if (when === undefined || when === null || when === "") return [];
  return [String(when)];
}

type WhenGate = { ok: true } | { ok: false; condition: string };

function evaluateGate(
  stepId: string,
  conditions: readonly string[],
  scope: Record<string, unknown>,
  warnings: string[],
): WhenGate {
  for (const condition of conditions) {
    let holds: boolean;
    try {
      holds = evaluateWhen(condition, scope);
    } catch (err) {
      // Never block a step on a typo: an unparseable condition is treated as
      // true but surfaced so the caller can report it.
      const message = err instanceof Error ? err.message : String(err);
      warnings.push(`when-unparseable:${stepId}:${message}`);
      continue;
    }
    if (!holds) return { ok: false, condition };
  }
  return { ok: true };
}

export function nextWave(def: WorkflowDef, state: State): Wave {
  const ready: Step[] = [];
  const skipped: SkipReason[] = [];
  const warnings: string[] = [];
  const scope = whenScope(state);

  for (const sdef of def.steps) {
    const st = state.steps[sdef.id];
    if (!st || st.status !== "pending") continue;
    const depIds = sdef.depends_on ?? [];
    const deps = depIds.flatMap((d) => {
      const dep = state.steps[d];
      return dep ? [{ id: d, status: dep.status }] : [];
    });
    const rule = sdef["trigger-rule"] ?? "all-success";
    const verdict = triggerRuleSatisfied(rule, deps);
    if (verdict.skip) {
      const summary = deps.map((d) => `${d.id}=${d.status}`).join(", ");
      skipped.push({ id: sdef.id, reason: `trigger-rule ${rule} not satisfied: ${summary}` });
      continue;
    }
    if (!verdict.runnable) continue;
    const gate = evaluateGate(sdef.id, whenConditions(sdef.when), scope, warnings);
    if (gate.ok) {
      ready.push(sdef);
    } else {
      skipped.push({ id: sdef.id, reason: `when: ${gate.condition} is false` });
    }
  }

  const statuses = Object.values(state.steps).map((s) => s.status);
  const inFlight = statuses.some((s) => s === "running");
  const anyPending = statuses.some((s) => s === "pending");
  const anyFailed = statuses.some((s) => s === "failed");
  const allDone = statuses.every((s) => TERMINAL_STEP.has(s));

  let done = false;
  let failed = false;
  if (ready.length === 0 && skipped.length === 0 && !inFlight) {
    if (anyFailed) {
      done = true;
      failed = true;
    } else if (allDone) {
      done = true;
    } else if (anyPending) {
      // Pending steps whose dependencies can never resolve: treat as failed.
      done = true;
      failed = true;
    }
  }
  return { ready, skipped, done, failed, warnings };
}

File: plugins/wise/engine/wise_engine/scheduler.py
from __future__ import annotations

import json
import math
import re
from collections.abc import Callable, Mapping, Sequence
from dataclasses import dataclass
from decimal import Decimal
from typing import Any

from .constants import TERMINAL_STEP, TRIGGER_RULES

UNKNOWN = object()
UNDEFINED = object()
_BARE_ROOTS = ("outputs", "inputs", "answers")
JS_WHITESPACE = "\t\n\v\f\r \u00a0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u2028\u2029\u202f\u205f\u3000\ufeff"
_SPACE = re.compile(f"[{re.escape(JS_WHITESPACE)}]")


def step_by_id(steps: Sequence[dict[str, Any]], identifier: str) -> dict[str, Any] | None:
    return next((step for step in steps if step["id"] == identifier), None)


def is_trigger_rule(rule: str) -> bool:
    return rule in TRIGGER_RULES


def trigger_rule_satisfied(rule: str, deps: Sequence[Mapping[str, Any]]) -> dict[str, bool]:
    if not deps:
        return {"runnable": True, "skip": False}
    statuses = [dep["status"] for dep in deps]
    completed = statuses.count("completed")
    failed = statuses.count("failed")
    all_terminal = all(status in TERMINAL_STEP for status in statuses)
    if rule == "one-success":
        return {"runnable": completed > 0, "skip": completed == 0 and all_terminal}
    if rule == "all-done":
        return {"runnable": all_terminal, "skip": False}
    if rule == "none-failed":
        return {"runnable": all_terminal and failed == 0, "skip": failed > 0}
    if rule == "none-failed-min-one-success":
        return {"runnable": all_terminal and failed == 0 and completed > 0, "skip": failed > 0}
    return {
        "runnable": completed == len(statuses),
        "skip": any(status in {"failed", "skipped", "cancelled"} for status in statuses),
    }


def number_text(value: int | float) -> str:
    number = float(value)
    if math.isnan(number):
        return "NaN"
    if math.isinf(number):
        return "Infinity" if number > 0 else "-Infinity"
    if number == 0:
        return "0"
    text = repr(number)
    if 1e-6 <= abs(number) < 1e21:
        decimal = format(Decimal(text), "f")
        return decimal.rstrip("0").rstrip(".") if "." in decimal else decimal
    mantissa, exponent = text.split("e")
    return f"{mantissa.removesuffix('.0')}e{int(exponent):+d}"


def js_string(value: Any) -> str:
    if value is UNDEFINED:
        return "undefined"
    if value is None:
        return "null"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return number_text(value)
    if isinstance(value, str):
        return value
    if isinstance(value, list):
        return ",".join(
            "" if item is None or item is UNDEFINED else js_string(item) for item in value
        )
    return "[object Object]"


def truthy(value: Any) -> bool:
    if value is UNDEFINED or value is None:
        return False
    if isinstance(value, str):
        return value != ""
    if isinstance(value, bool):
        return value
    if isinstance(value, (int, float)):
        return value != 0 and not math.isnan(value)
    if isinstance(value, list):
        return len(value) > 0
    return True


def _values_equal(left: Any, right: Any) -> bool:
    if left is right:
        return True
    if left is UNDEFINED or right is UNDEFINED or left is None or right is None:
        return False
    if isinstance(left, (dict, list)) or isinstance(right, (dict, list)):
        return False
    return js_string(left) == js_string(right)


def _field(value: Any, key: str) -> Any:
    if isinstance(value, Mapping):
        return value.get(key, UNDEFINED)
    if isinstance(value, list) and key == "length":
        return len(value)
    return UNDEFINED


def _owner_of(head: str, scope: Mapping[str, Any]) -> Any:
    if head in scope:
        return scope
    for root in _BARE_ROOTS:
        bucket = scope.get(root)
        if _field(bucket, head) is not UNDEFINED:
            return bucket
        if isinstance(bucket, Mapping) and head in bucket:
            return bucket
    return UNDEFINED


def resolve_identifier(name: str, scope: Mapping[str, Any]) -> Any:
    path = name.split(".")
    value = _owner_of(path[0], scope)
    for segment in path:
        value = _field(value, segment)
    return value


def resolve_identifier_partial(name: str, scope: Mapping[str, Any]) -> Any:
    path = name.split(".")
    value = _owner_of(path[0], scope)
    if value is UNDEFINED:
        return UNKNOWN
    for segment in path:
        if isinstance(value, Mapping) and segment in value:
            value = value[segment]
        elif isinstance(value, list) and segment == "length":
            value = len(value)
        else:
            return UNKNOWN
    return value


def _when_error(expr: str, message: str, position: int) -> ValueError:
    offset = len(expr[:position].encode("utf-16-le", errors="surrogatepass")) // 2
    return ValueError(
        f"when: {message} at position {offset} in {json.dumps(expr, ensure_ascii=False)}"
    )


@dataclass(frozen=True)
class _Token:
    kind: str
    value: str
    position: int


def _tokenize(expr: str) -> list[_Token]:
    tokens = []
    position = 0
    while position < len(expr):
        char = expr[position]
        start = position
        if _SPACE.fullmatch(char):
            position += 1
            continue
        if char in "()":
            tokens.append(_Token("lparen" if char == "(" else "rparen", char, start))
            position += 1
        elif char in "\"'":
            end = expr.find(char, position + 1)
            if end < 0:
                raise _when_error(expr, "unterminated string", start)
            tokens.append(_Token("string", expr[position + 1 : end], start))
            position = end + 1
        elif "0" <= char <= "9":
            end = position
            while end < len(expr) and expr[end] in "0123456789.":
                end += 1
            text = expr[position:end]
            if re.fullmatch(r"[0-9]+(\.[0-9]+)?", text) is None:
                raise _when_error(expr, f"bad number {text}", start)
            tokens.append(_Token("number", text, start))
            position = end
        elif re.fullmatch(r"[A-Za-z_]", char):
            end = position
            while end < len(expr) and re.fullmatch(r"[A-Za-z0-9_.]", expr[end]):
                end += 1
            text = expr[position:end]
            if re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*(\.[A-Za-z_][A-Za-z0-9_]*)*", text) is None:
                raise _when_error(expr, f"bad identifier {text}", start)
            tokens.append(_Token("ident", text, start))
            position = end
        else:
            two = expr[position : position + 2]
            if two in {"==", "!=", "&&", "||"}:
                tokens.append(_Token("op", two, start))
                position += 2
            elif char == "!":
                tokens.append(_Token("op", char, start))
                position += 1
            else:
                encoded = json.dumps(char, ensure_ascii=False)
                if ord(char) > 0xFFFF:
                    high = char.encode("utf-16-le")[:2].decode("utf-16-le", errors="surrogatepass")
                    encoded = json.dumps(high)
                raise _when_error(expr, f"unexpected character {encoded}", start)
    tokens.append(_Token("end", "", len(expr)))
    return tokens


def _truthy3(value: Any) -> bool | None:
    return None if value is UNKNOWN else truthy(value)


class _Evaluator:
    def __init__(self, expr: str, resolve: Callable[[str], Any]) -> None:
        self.expr = expr
        self.resolve = resolve
        self.tokens = _tokenize(expr)
        self.index = 0

    def peek(self) -> _Token:
        return self.tokens[self.index]

    def take(self) -> _Token:
        token = self.peek()
        self.index += 1
        return token

    def is_op(self, value: str) -> bool:
        return self.peek().kind == "op" and self.peek().value == value

    def primary(self) -> Any:
        token = self.take()
        if token.kind == "lparen":
            value = self.parse_or()
            if self.peek().kind != "rparen":
                raise _when_error(self.expr, "expected ')'", self.peek().position)
            self.index += 1
            return value
        if token.kind == "string":
            return token.value
        if token.kind == "number":
            return float(token.value)
        if token.kind == "ident":
            if token.value in {"true", "false"}:
                return token.value == "true"
            return self.resolve(token.value)
        if token.kind == "end":
            raise _when_error(self.expr, "unexpected end of expression", token.position)
        raise _when_error(self.expr, f"unexpected {json.dumps(token.value)}", token.position)

    def unary(self) -> Any:
        if self.is_op("!"):
            self.index += 1
            value = _truthy3(self.unary())
            return UNKNOWN if value is None else not value
        return self.primary()

    def equality(self) -> Any:
        left = self.unary()
        while self.is_op("==") or self.is_op("!="):
            operation = self.take().value
            right = self.unary()
            if left is UNKNOWN or right is UNKNOWN:
                left = UNKNOWN
            else:
                equal = _values_equal(left, right)
                left = equal if operation == "==" else not equal
        return left

    def parse_and(self) -> Any:
        left = self.equality()
        while self.is_op("&&"):
            self.index += 1
            lhs, rhs = _truthy3(left), _truthy3(self.equality())
            left = (
                False
                if lhs is False or rhs is False
                else (True if lhs is True and rhs is True else UNKNOWN)
            )
        return left

    def parse_or(self) -> Any:
        left = self.parse_and()
        while self.is_op("||"):
            self.index += 1
            lhs, rhs = _truthy3(left), _truthy3(self.parse_and())
            left = (
                True
                if lhs is True or rhs is True
                else (False if lhs is False and rhs is False else UNKNOWN)
            )
        return left

    def evaluate(self) -> bool | None:
        result = self.parse_or()
        tail = self.peek()
        if tail.kind != "end":
            raise _when_error(self.expr, f"unexpected {json.dumps(tail.value)}", tail.position)
        return _truthy3(result)


def evaluate_when(expr: str, scope: Mapping[str, Any]) -> bool:
    return _Evaluator(expr, lambda name: resolve_identifier(name, scope)).evaluate() is True


def evaluate_when_partial(expr: str, scope: Mapping[str, Any]) -> bool | None:
    return _Evaluator(expr, lambda name: resolve_identifier_partial(name, scope)).evaluate()


def when_scope(state: Mapping[str, Any]) -> dict[str, Any]:
    return {key: state.get(key, UNDEFINED) for key in _BARE_ROOTS}


def when_conditions(when: Any = UNDEFINED) -> list[str]:
    if isinstance(when, list):
        return [js_string(condition) for condition in when]
    if when is UNDEFINED or when is None or when == "":
        return []
    return [js_string(when)]


def next_wave(definition: Mapping[str, Any], state: Mapping[str, Any]) -> dict[str, Any]:
    ready, skipped, warnings = [], [], []
    scope = when_scope(state)
    for step in definition["steps"]:
        current = state["steps"].get(step["id"])
        if current is None or current.get("status") != "pending":
            continue
        deps = [
            {"id": key, "status": state["steps"][key].get("status", UNDEFINED)}
            for key in (step.get("depends_on") or [])
            if state["steps"].get(key) is not None
        ]
        rule = step.get("trigger-rule")
        if rule is None:
            rule = "all-success"
        verdict = trigger_rule_satisfied(rule, deps)
        if verdict["skip"]:
            summary = ", ".join(f"{dep['id']}={dep['status']}" for dep in deps)
            skipped.append(
                {"id": step["id"], "reason": f"trigger-rule {rule} not satisfied: {summary}"}
            )
            continue
        if not verdict["runnable"]:
            continue
        for condition in when_conditions(step.get("when")):
            try:
                holds = evaluate_when(condition, scope)
            except ValueError as error:
                warnings.append(f"when-unparseable:{step['id']}:{error}")
                continue
            if not holds:
                skipped.append({"id": step["id"], "reason": f"when: {condition} is false"})
                break
        else:
            ready.append(step)
    statuses = [step.get("status", UNDEFINED) for step in state["steps"].values()]
    done = failed = False
    if not ready and not skipped and "running" not in statuses:
        if "failed" in statuses:
            done = failed = True
        elif all(status in TERMINAL_STEP for status in statuses):
            done = True
        elif "pending" in statuses:
            done = failed = True
    return {
        "ready": ready,
        "skipped": skipped,
        "done": done,
        "failed": failed,
        "warnings": warnings,
    }
